### ISSUE TITLE
Ziped Files and Descriptors not usable for some workflows 

### ISSUE THREAD

**Issue by patmagee** [dockstore/dockstore#5594]
**Describe the bug**
(Preface that I have only looked at WDL workflows)

Many workflows in dockstore come from repositories that have complex structures where the primary descriptor actually imports workflows in other parent directories.

For example:

```
workflows
  |- wgs
    | - main.wdl
tasks
  |- do-stuff.wdl
structs
  | - human
    | - sample.wdl
```

When dockstore imports these workflows it assumes that the PRIMARY_DESCRIPTOR is at the root of the file tree and strips any leading path away, resolving all imports against the primary descriptor. This works when all imported files are siblings o of the primary descriptor or in child directories. This completely FAILS when the imports are in a totally separate parent directory. Since they are resolved relative to the root primary file their path ends up being prepended by back references `../../` instead of keeping the leading paths between `main.wdl` and `do-stuff.wdl`

```
main.wdl
../../tasks/do-stuff.wdl
../../structs/human/sample.wdl
```

**So what is the big deal??**

The problem with this is that workflows which have this structure are **NOT RUNNABLE** when passing the `PLAIN_WDL/descriptor` to any workflow engine (ie miniwdl or cromwell) they will not be able to resolve the workflow.

Additionally when trying to download the workflow files as a zip, the dockstore server is unable to create the zip file likely due to the fact that it cannot create paths with back references as zip entries 

- - -
**To Reproduce**

Workflow ID: '#workflow/github.com/broadinstitute/warp/WholeGenomeGermlineSingleSample'
Version ID: `tw_GL-2036_create_rtools_docker`
Type: `WDL`

1. Got to the [Swagger](https://dockstore.org/api/static/swagger-ui/index.html#/GA4GHV20/toolsIdVersionsVersionIdTypeFilesGet) uI and navigate to the TRS request to get the files
1. Try out the request
1. Insert the above `Workflow ID` in the workflow id input
1. Insert the above `Version ID` in the version id input
1. Set the type to `WDL` and the `format` to `zip`
1. Click execute

**Expected behavior**
The Zip file to be downloaded with all of the files for the workflow in it

**Actual Behaviour**

```
{
  "servlet": "jersey",
  "message": "Bad Request",
  "url": "/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/WDL/files",
  "status": "400"
}
```

- - -
**To Reproduce**

Workflow ID: '#workflow/github.com/broadinstitute/warp/WholeGenomeGermlineSingleSample'
Version ID: `tw_GL-2036_create_rtools_docker`
Type: `WDL`
Descriptor URL: `'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor`

1 . Try to run the workflow using miniwdl (without passing any inputs)

```
miniwdl run https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor
```

**Expected behavior**
The workflow to be parsed and attempted to run (but fail if you do not provide inputs)

**Actual Behaviour**

```
miniwdl run 'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor'
(https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor Ln 31 Col 1) Failed to import ../../../../../../tasks/broad/UnmappedBamToAlignedBam.wdl
HTTP Error 404: Not Found
```

- - -
**To Reproduce**

Workflow ID: '#workflow/github.com/broadinstitute/warp/WholeGenomeGermlineSingleSample'
Version ID: `tw_GL-2036_create_rtools_docker`
Type: `WDL`
Descriptor URL: `'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor`

1. Start cromwell: `java -jar cromwell.jar server`
1. Try to run the workflow using cromwell (without passing any inputs)

```
curl -X POST http://localhost:8000/api/workflows/v1 -F workflowSource='https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor'
```

**Expected behavior**
The workflow to be parsed and attempted to run (but fail if you do not provide inputs)

**Actual Behaviour**
Cromwell is not able to parse the workflow imports

```
[2023-07-12 11:01:23,90] [info] WorkflowManagerActor: Workflow 9b777f37-fa53-490b-9478-fd4d96b9a92c failed (during MaterializingWorkflowDescriptorState): cromwell.engine.workflow.lifecycle.materialization.MaterializeWorkflowDescriptorActor$$anon$1: Workflow input processing failed:
Failed to import '../../../../../../tasks/broad/UnmappedBamToAlignedBam.wdl' (reason 1 of 1): Failed to resolve '../../../../../../tasks/broad/UnmappedBamToAlignedBam.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/tasks/broad/UnmappedBamToAlignedBam.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
Failed to import '../../../../../../tasks/broad/AggregatedBamQC.wdl' (reason 1 of 1): Failed to resolve '../../../../../../tasks/broad/AggregatedBamQC.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/tasks/broad/AggregatedBamQC.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
Failed to import '../../../../../../tasks/broad/Qc.wdl' (reason 1 of 1): Failed to resolve '../../../../../../tasks/broad/Qc.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/tasks/broad/Qc.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
Failed to import '../../../../../../tasks/broad/BamToCram.wdl' (reason 1 of 1): Failed to resolve '../../../../../../tasks/broad/BamToCram.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/tasks/broad/BamToCram.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
Failed to import '../../../../../../tasks/broad/Utilities.wdl' (reason 1 of 1): Failed to resolve '../../../../../../tasks/broad/Utilities.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/tasks/broad/Utilities.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
Failed to import '../../../../../../pipelines/broad/dna_seq/germline/variant_calling/VariantCalling.wdl' (reason 1 of 1): Failed to resolve '../../../../../../pipelines/broad/dna_seq/germline/variant_calling/VariantCalling.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/pipelines/broad/dna_seq/germline/variant_calling/VariantCalling.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
Failed to import '../../../../../../structs/dna_seq/DNASeqStructs.wdl' (reason 1 of 1): Failed to resolve '../../../../../../structs/dna_seq/DNASeqStructs.wdl' using resolver: 'http importer (relative to https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/)' (reason 1 of 1): Failed to download https://dockstore.org/api/ga4gh/trs/structs/dna_seq/DNASeqStructs.wdl (reason 1 of 1): {"code":404,"message":"HTTP 404 Not Found"}
	at cromwell.engine.workflow.lifecycle.materialization.MaterializeWorkflowDescriptorActor.cromwell$engine$workflow$lifecycle$materialization$MaterializeWorkflowDescriptorActor$$workflowInitializationFailed(MaterializeWorkflowDescriptorActor.scala:257)
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/DOCK-2430)
┆Fix Versions: Dockstore 1.16
┆Issue Number: DOCK-2430
┆Sprint: 117 - Pearl
┆Issue Type: Story

**Comment by patmagee** [comment]
See 5182

**Comment by denis-yuen** [comment]
Some thoughts from DMs

>  Probably affects CWL too, but not Nextflow (think it assumes the config is in the root)
>  We were kinda ambiguous on the format of the zip file. We could return a zip file starting at the "lowest" point of the data structure. We'd still need to return a header or somehow annotate/metadata the zip file to indicate which wdl was the primary descriptor. (or just correlate the zip and the json return)

> The only way I can think of to fix resolving these types of imports against the /decriptor endpoint in TRS is to either:
> Assume the client understands how to treat these strings in a special way, different then normal URL resolution semantics
require files to either be siblings or children of the primary descriptor

**Comment by denis-yuen** [comment]
Looking specifically at the zip issue

> Additionally when trying to download the workflow files as a zip, the dockstore server is unable to create the zip file likely due to the fact that it cannot create paths with back references as zip entries

This looks like a separate issue, basically, you need to make sure the media type matches the `format` or only use the media type

Or in curl terms 

Right
```
curl -X 'GET' \
  'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/WDL/files?format=zip' \
  -H 'accept: application/zip'
```

```
curl -X 'GET' \
  'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/WDL/files' \
  -H 'accept: application/zip'
```

Wrong (gets 400)
```
curl -X 'GET' \
  'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/WDL/files?format=zip' \
  -H 'accept: application/json'
```

**Comment by denis-yuen** [comment]
Alternatively, the edit to the instructions is 

1. Go to the [Swagger](https://dockstore.org/api/static/swagger-ui/index.html#/GA4GHV20/toolsIdVersionsVersionIdTypeFilesGet) uI and navigate to the TRS request to get the files
1. Try out the request
1. Insert the above `Workflow ID` in the workflow id input
1. Insert the above `Version ID` in the version id input
1. Set the type to `WDL` and the `format` to `zip`
1. **Change the media type to application/zip**
1. Click execute

or 

1. Got to the [Swagger](https://dockstore.org/api/static/swagger-ui/index.html#/GA4GHV20/toolsIdVersionsVersionIdTypeFilesGet) uI and navigate to the TRS request to get the files
1. Try out the request
1. Insert the above `Workflow ID` in the workflow id input
1. Insert the above `Version ID` in the version id input
1. Set the type to `WDL`
1. Change the media type to application/zip
1. Click execute

**Comment by denis-yuen** [comment]
Interestingly 

> **Actual Behaviour**
> 
> ```
> miniwdl run 'https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor'
> (https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor Ln 31 Col 1) Failed to import ../../../../../../tasks/broad/UnmappedBamToAlignedBam.wdl
> HTTP Error 404: Not Found
> ```

https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/WDL/descriptor/..%2F..%2F..%2F..%2F..%2F..%2Ftasks%2Fbroad%2FUnmappedBamToAlignedBam.wdl is a live link 

Thinking miniwdl must be collapsing the `../` parts

**Comment by denis-yuen** [comment]
Tinkering with miniwdl

```
$ miniwdl run https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor
before:https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor../../../../../../tasks/broad/UnmappedBamToAlignedBam.wdl
after:https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor/..%2F..%2F..%2F..%2F..%2F..%2Ftasks%2Fbroad%2FUnmappedBamToAlignedBam.wdl
before:https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor/..%2F..%2F..%2F..%2F..%2F..%2Ftasks%2Fbroad%2FUnmappedBamToAlignedBam.wdl../../tasks/broad/Alignment.wdl
after:https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor/..%2F..%2F..%2F..%2F..%2F..%2Ftasks%2Fbroad%2FUnmappedBamToAlignedBam.wdl/..%2F..%2Ftasks%2Fbroad%2FAlignment.wdl
(https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor Ln 31 Col 1) Failed to import ../../../../../../tasks/broad/UnmappedBamToAlignedBa
![](https://media.tenor.com/Ndd63HyG3KcAAAAd/julia-louis-dreyfus-nervous-laughter.gif)
m.wdl
(../../../../../../tasks/broad/UnmappedBamToAlignedBam.wdl Ln 19 Col 1) Failed to import ../../tasks/broad/Alignment.wdl
HTTP Error 404: Not Found : version found, but file not found (bad filename, invalid file, etc.)
```
![julia-louis-dreyfus-nervous-laughter](https://github.com/dockstore/dockstore/assets/1730679/d77e311f-7115-48c9-8f9c-e3a324452dd9)


Particularly
```
https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fbroadinstitute%2Fwarp%2FWholeGenomeGermlineSingleSample/versions/tw_GL-2036_create_rtools_docker/PLAIN_WDL/descriptor/..%2F..%2F..%2F..%2F..%2F..%2Ftasks%2Fbroad%2FUnmappedBamToAlignedBam.wdl/..%2F..%2Ftasks%2Fbroad%2FAlignment.wdl
```


Uh oh
